### PR TITLE
Restrict PartialOrd comparison to only tokens in the same buffer

### PR DESCRIPTION
**Pull request by dtolnay**
How does this look @CAD97? I was hesitant about directly exposing a pointer comparison across unrelated allocations.

```rust
use syn::buffer::TokenBuffer;
let a = TokenBuffer::new2("...".parse().unwrap());
let b = TokenBuffer::new2("...".parse().unwrap());
println!("{:?}", a.begin().partial_cmp(&b.begin()));
```

### CODE REVIEW

**Review by CAD97 — COMMENTED** [review]
That looks really good to me! It's definitely nice to have the cursors for distinct buffers unordered rather than arbitrary-but-consistent order.

It's nice it also gives a purpose to the `End` entries rather than them solely existing to make implementation easier.

(And as a tertiary nicety, it doesn't necessarily preclude a rewrite of `Cursor` that uses slices and less `unsafe`, as such could still do the offset for comparison, just not access outside the cursor range.)

TLDR: r+

(Apparently, "Can not approve pull request without explicit repository access" — did GH decide to prevent drive-by PR approvals? Or is that an old setting I don't know about? Irrelevant, I guess; though it is nice to be able to give a structured ±1 without merge capabilities, it's definitely abusable.)
